packages/medusa/src/api/utils/middlewares/products/filter-by-valid-sales-channels.ts

```diff
@@ -1,79 +1,74 @@
 
-    if (!store) {
-      try {
-        throw new MedusaError(MedusaError.Types.INVALID_DATA, `Store not found`)
-      } catch (e) {
-        return next(e)
+      if (!apiKey) {
+        return next(
+          new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `Publishable API key not found`
+          )
+        )
+      }
+      let result = apiKey.sales_channels_link.map(
+        (link) => link.sales_channel_id
+      )
+
+      if (salesChannelIds?.length) {
+        // If all sales channel ids are not in the publishable key, we throw an error
+        const uniqueInParams = arrayDifference(salesChannelIds, result)
+        if (uniqueInParams.length) {
+          return next(
+            new MedusaError(
+              MedusaError.Types.INVALID_DATA,
+              `Requested sales channel is not part of the publishable key mappings`
+            )
+          )
+        }
+        result = salesChannelIds
       }
-    }
-    // Always set sales channels in the following priority
-    // - any existing sales chennel ids passed through query params
-    // - if none, we set the default sales channel
-    req.filterableFields.sales_channel_id = salesChannelIds ?? [
-      store.default_sales_channel_id,
-    ]
 
-    // Return early if no publishable keys are found
-    if (!isPresent(publishableApiKey)) {
+      req.filterableFields.sales_channel_id = result
       return next()
     }
 
-    // When publishable keys are present, we fetch for all sales chennels attached
-    // to the publishable key and validate the sales channel filter against it
-    const apiKey = await refetchEntity(
-      "api_key",
-      { token: publishableApiKey },
-      req.scope,
-      ["id", "sales_channels_link.sales_channel_id"]
-    )
-
-    if (!apiKey) {
-      try {
-        throw new MedusaError(
-          MedusaError.Types.INVALID_DATA,
-          `Publishable API key not found`
-        )
-      } catch (e) {
-        return next(e)
-      }
+    if (salesChannelIds) {
+      req.filterableFields.sales_channel_id = salesChannelIds
+      return next()
     }
 
-    const validSalesChannelIds = apiKey.sales_channels_link.map(
-      (link) => link.sales_channel_id
-    )
-
-    const invalidSalesChannelIds = (salesChannelIds || []).filter(
-      (id) => !validSalesChannelIds.includes(id)
-    )
+    const store = await refetchEntity("stores", {}, req.scope, [
+      "default_sales_channel_id",
+    ])
 
-    if (invalidSalesChannelIds.length) {
-      try {
-        throw new MedusaError(
-          MedusaError.Types.INVALID_DATA,
-          `Invalid sales channel filters provided - ${invalidSalesChannelIds.join(
-            ", "
-          )}`
-        )
-      } catch (e) {
-        return next(e)
-      }
+    if (!store) {
+      return next(
+        new MedusaError(MedusaError.Types.INVALID_DATA, `Store not found`)
+      )
     }
 
+    req.filterableFields.sales_channel_id = [store.default_sales_channel_id]
```

Comment:
sugg: this could still be undefined, might have to throw here in those cases

Reply:
Good point, I guess there is no guarantee that there is a default sales channel